boards/components/src/console.rs

```diff
@@ -74,48 +97,57 @@ impl Component for UartMuxComponent {
     }
 }
 
-pub struct ConsoleComponent {
+#[macro_export]
+macro_rules! console_component_helper {
+    () => {{
+        use capsules::console::{Console, BUF_SIZE};
+        use kernel::static_buf;
+        let read_buf = static_buf!([u8; BUF_SIZE]);
+        let write_buf = static_buf!([u8; BUF_SIZE]);
+        let console = static_buf!(Console<'static>);
+        (write_buf, read_buf, console)
+    }};
+}
+
+pub struct ConsoleComponent<T: 'static + hil::uart::UartData<'static>> {
     board_kernel: &'static kernel::Kernel,
     driver_num: usize,
-    uart_mux: &'static MuxUart<'static>,
+    uart: &'static T,
 }
 
-impl ConsoleComponent {
-    pub fn new(
-        board_kernel: &'static kernel::Kernel,
-        driver_num: usize,
-        uart_mux: &'static MuxUart,
-    ) -> ConsoleComponent {
+impl<T: hil::uart::UartData<'static>> ConsoleComponent<T> {
+    pub fn new(board_kernel: &'static kernel::Kernel, driver_num: usize, uart: &'static T) -> Self {
         ConsoleComponent {
             board_kernel: board_kernel,
             driver_num: driver_num,
-            uart_mux: uart_mux,
+            uart,
         }
     }
 }
 
-impl Component for ConsoleComponent {
-    type StaticInput = ();
+impl<T: hil::uart::UartData<'static>> Component for ConsoleComponent<T> {
+    type StaticInput = (
+        StaticUninitializedBuffer<[u8; BUF_SIZE]>,
+        StaticUninitializedBuffer<[u8; BUF_SIZE]>,
+        StaticUninitializedBuffer<console::Console<'static>>,
+    );
     type Output = &'static console::Console<'static>;
 
-    unsafe fn finalize(self, _s: Self::StaticInput) -> Self::Output {
+    unsafe fn finalize(self, s: Self::StaticInput) -> Self::Output {
         let grant_cap = create_capability!(capabilities::MemoryAllocationCapability);
 
-        // Create virtual device for console.
-        let console_uart = static_init!(UartDevice, UartDevice::new(self.uart_mux, true));
-        console_uart.setup();
```

Comment:
I am not sure I agree with moving the creation of the virtual device outside of the component. The purpose of components is to enable simple, single-line initialization of capsules for the common case, to reduce code duplication across our many upstream boards. It makes sense to not use a mux for something like GNSS, but that is not a  common use case and could just be initialized without using the component at all. We have run into many users doing things like forgetting to call `setup()` on the virtual device and that being included in the component prevents that error.

I think we should either have a `NonMuxedConsoleComponent` to support things like GNSS, or just have that use case go without a component.

Reply:
I brought it back to mux only.